[package] issue error on dunder import failures

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59022 [package] issue warning instead of error for dunder imports failures**

When a user has a dependency that uses `__import__` with dynamic arguments, issue an error when we fail to import it.

Example trigger: 
```
import torch
from textwrap import dedent

src = dedent(
                """\
                fake_name = "time"
                __import__(fake_name) # need to throw error

                def foo(mod_name: str):
                    __import__(mod_name) # need to not throw error? this PR will flag this as an error .-.
                """
            )

with torch.package.PackageExporter("output", verbose=False) as e:
    e.save_source_string('broken',  src)
```
Output: 
```
Traceback (most recent call last):
  File "ztest_space.py", line 15, in <module>
    e.save_source_string('broken',  src)
  File "/home/lillianjohnson/local/pytorch/torch/package/package_exporter.py", line 750, in __exit__
    self.close()
  File "/home/lillianjohnson/local/pytorch/torch/package/package_exporter.py", line 843, in close
    self._execute_dependency_graph()
  File "/home/lillianjohnson/local/pytorch/torch/package/package_exporter.py", line 779, in _execute_dependency_graph
    self._validate_dependency_graph()
  File "/home/lillianjohnson/local/pytorch/torch/package/package_exporter.py", line 766, in _validate_dependency_graph
    raise PackagingError(self.dependency_graph)
torch.package.package_exporter.PackagingError: 
* Dependency resolution failed.
    broken
      Context: Error encountered in determining the target dependency for an __import__ call in module 'broken', lineno: 2. Support for __import__ is limited to constant arguments. Error encountered: 'Name' object has no attribute 'value'.
```

Differential Revision: [D28723825](https://our.internmc.facebook.com/intern/diff/D28723825)

Commit:
[package] issue warning instead of error for dunder imports failures

[ghstack-poisoned]

Lilyjjo commented:
I question if I should gate the warning being printed to just when the `PackageExporter`'s verbosity is set to `True`. I get annoyed when I don't know how to silence warnings

Commit:
Update on "[package] issue warning instead of error for dunder imports failures"


When a user has a dependency that uses `__import__` with dynamic arguments, issue a warning and skip the import instead of failing. This PR just wraps the `__import__` logic in a try-catch block. 

```
import torch

with torch.package.PackageExporter("output", verbose=False) as e:
    e.save_source_string('m',  '__import__("these", dont, have, to, be, contants)') # can't resolve dynamic arguments
```
output: 
```
/pytorch/torch/package/find_file_dependencies.py:103: UserWarning: Error importing dynamic __import__ from module 'm', lineno: 1. torch.package does not support dynamic imports. Error encountered: 'Name' object has no attribute 'elts'.
  warnings.warn(
```


[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28723825).

Commit:
Update on "[package] issue warning instead of error for dunder imports failures"


When a user has a dependency that uses `__import__` with dynamic arguments, issue a warning and skip the import instead of failing. This PR just wraps the `__import__` logic in a try-except block. 

Example trigger: 
```
import torch

with torch.package.PackageExporter("output", verbose=False) as e:
    e.save_source_string('m',  '__import__("these", dont, have, to, be, contants)') # can't resolve dynamic arguments
```
Output: 
```
/pytorch/torch/package/find_file_dependencies.py:103: UserWarning: PackageExporter encountered an error in determining the dependency for an __import__ call in module 'm', lineno: 1; skipping this dependency. Dynamic __import__ support is not implemented. Error encountered: 'Name' object has no attribute 'elts'.
  warnings.warn(
```

Differential Revision: [D28723825](https://our.internmc.facebook.com/intern/diff/D28723825)

[ghstack-poisoned]
